### ISSUE TITLE
Username with German letters gets encoded and passes validation ASP.NET Core Identity

### ISSUE THREAD

**Issue by Ponant** [aspnet/AspNetCore#9107]
On .Net Core 3.0 Preview 3, Asp.NET Core Identity Template, register a user with username
`bernhard@günter.com
`
**Expected**
Validation throws that this is not an allowed character for a username since the default is
`
public string AllowedUserNameCharacters { get; set; }
            = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-._@+";
`
**Result**
`bernhard@xn--gnter-kva.com 
` is stored in the db as username and email

That is pretty unexpected for an old framework and a handicap in the current implementation.
Probably this has been encountered in the past??

**Comment by Ponant** [comment]
A first issue comes from the template having the `[EmailAddress]` attribute. If that is commented out then `günter` is not formatted and the response is as expected "User name 'bernhard@günter.com' is invalid, can only contain letters or digits.", which btw is a wrong error description since other characters are allowed by default (-._@+) .
The problem remains when someone wants to collect an email address and puts in the `[EmailAddress]` he will collect a wrong email which can have lost of consequences.

**Comment by Ponant** [comment]
The same issue happens with `[DataType(DataType.EmailAddress)]`

**Comment by blowdart** [comment]
That's odd. @HaoK ?

**Comment by HaoK** [comment]
There's two kinds of validation going on, the MVC/model data validation attributes which control the views/UI validation on the forms, and then there's a set of identity specific properties that do things inside of the identity methods.

The templates use email as the username and set it as so, which is why that happens.  It also does it in two steps, it updates the email ( which works), and then tries to update the username which fails.

The relevant code which clearly can result in out of sync username/emails:
```
                var setEmailResult = await _userManager.SetEmailAsync(user, Input.Email);
                if (!setEmailResult.Succeeded)
                {
                    var userId = await _userManager.GetUserIdAsync(user);
                    throw new InvalidOperationException($"Unexpected error occurred setting email for user with ID '{userId}'.");
                }

                // In our UI email and user name are one and the same, so when we update the email
                // we need to update the user name.
                var setUserNameResult = await _userManager.SetUserNameAsync(user, Input.Email);
                if (!setUserNameResult.Succeeded)
                {
                    var userId = await _userManager.GetUserIdAsync(user);
                    throw new InvalidOperationException($"Unexpected error occurred setting name for user with ID '{userId}'.");
                }
```

**Comment by Ponant** [comment]
I disagree with the arguments.
1) The `[DataType(DataType.EmailAddress)] ` or `[EmailAddress]` will encode the umlaut character _ü_ silently and transform it somewhat to some combination of dashes and dots and some alphanumeric (see above), and that will end up in the database and ui for both username and email since now the former is encoded with dashes and the like, so it passes validation, see:
```
   private async Task ValidateUserName(UserManager<TUser> manager, TUser user,
            ICollection<IdentityError> errors)
```
A proper attribute or whatever logic would stop me from sending an encoded umlaut to the server. Somehow we should make sure that the user input is properly sent to the db and later the ui.

2) This violation occurs when the umlaut or a French accent is in the domain name. güter@gmail.com would not pass the validation on the client, even if js validation is turned off

2) A round trip to the server is needed to validate the UserName, I am now working on having something validating client-side following the rules of `AllowedCharacters`

3) The `InvalidUserName` key the resx file has a value which is not consistent with the default values of the `AllowedCharacters` of the `identityOptions`

It would be good to have feedback from some people who I think are used to German, e.g.
@leastprivilege or French @PinpointTownes

**Comment by PinpointTownes** [comment]
> The [DataType(DataType.EmailAddress)] or [EmailAddress] will encode the umlaut character ü silently and transform it somewhat to some combination of dashes and dots and some alphanumeric (see above)

It's actually called [punycode](https://en.wikipedia.org/wiki/Punycode).

> It would be good to have feedback from some people who I think are used to German, e.g.
@leastprivilege or French @PinpointTownes

That's a rather strange bug, indeed (and not something I've seen before, tho' I must admit I don't use accents in my email addresses, which is often not allowed by major email providers anyway).

Are you sure the username/email are not punycoded client-side by the browser before being sent to the server? It may explain why the validation check doesn't reject the punycoded username/email if they are received already encoded.

**Comment by Ponant** [comment]
@PinpointTownes  I will do more checks later as it is late here; all I can say for now is that if you remove I will receive punycoded data only if I have these EmailAddress attributes. This holds with or without validation js scripts. Otherwise the data will go with the umlaut to the server and UserValidation will reject the username (as expected). Does this tell you something?

**Comment by Ponant** [comment]
I think this should be moved to corefx as it does not seem to be related to Identity. I will update this thread once the bug is isolated.
https://github.com/dotnet/corefx/issues/36675

**Comment by PinpointTownes** [comment]
> I will do more checks later as it is late here; all I can say for now is that I will receive punycoded data only if I have these EmailAddress attributes. Does this tell you something?

These attributes affect the way email input boxes are represented in HTML5 (i.e as `type=email` inputs rather than `type=text`).

A quick search shows that it is indeed a browser-side conversion, that automatically happens for email fields: https://bugs.chromium.org/p/chromium/issues/detail?id=410937

It's also interesting to note that the W3C HTML WG updated the spec to [allow non-ASCII chars in email inputs](https://github.com/w3c/html/issues/845). I wouldn't be surprised if the behavior you're seeing was changed in a future version of mainstream browsers.

**Comment by Ponant** [comment]
@PinpointTownes , then why I do not have this problem with a fresh non core asp.net identity but still on chrome?

**Comment by Ponant** [comment]
@PinpointTownes , ok I just found out that the old .Net framework outputs type=text while .Net Core type=email. I also found that FF does send the data correctly to the browser, so it seems to be a chromium issue.

**Comment by Ponant** [comment]
the bug you link to dates back to 2014, so I still think  that should be fixed here to accommodate for chrome since the response from chromium has been a "won't fix".

**Comment by PinpointTownes** [comment]
> ok I just found out that the old .Net framework outputs type=text while .Net Core type=email. I also found that FF does send the data correctly to the browser, so it seems to be a chromium issue.

Yep, that's because the MVC 5.x templates use `Html.TextBoxFor` for the email/username field. If you switch to `Html.EditorFor`, it will take the `EmailAddress` attribute into account and you'll see a similar behavior.

> the bug you link to dates back to 2014, so I still think that should be fixed here to accommodate for chrome since the response from chromium has been a "won't fix".

At best, Identity's UI could try to proactively "punydecode" the username/email fields back to their real Unicode representation and validate them after that. Whether it's a good idea or something rather fragile is something for @blowdart to determine 😄

**Comment by Ponant** [comment]
Yes, well that is maybe an issue for the  `asp-for` tag helper

**Comment by blowdart** [comment]
Given there's RFCs for international characters in email addresses now, (RFCs: 6530, 6531,6532, 6533), and even Office 365 supports it (and let's face it, Microsoft isn't the fastest to jump on board to new standards), allowing international characters is probably a valid choice. Chrome's conversion however, in my view, isn't. 

If anything I'd lean to the "un-punycode", but that would have to be tested a bunch to ensure "fake" punycode isn't going to cause crashes, and assumes there's an un-punycode in CoreFX somewhere.

**Comment by PinpointTownes** [comment]
> and assumes there's an un-punycode in CoreFX somewhere

There's [IdnMapping](https://docs.microsoft.com/en-us/dotnet/api/system.globalization.idnmapping?view=netframework-4.7.2) (already used in ASP.NET Core), but it's completely specific to domain names and I'm not sure it would work with email addresses.

**Comment by Ponant** [comment]
Why not reverting back to type=text, it works in that I avoid formatting surprise and then it is up to the dev to decide if he wants non us-ascii or not?

**Comment by PinpointTownes** [comment]
Well, that would mean degrading the UX a bit, as `type=email` inputs allow mobile browsers to render a keyboard with everything you need to type an email address (typically, the `@`). It's not exactly something you can blindly change without weighing the pros/cons.

**Comment by Ponant** [comment]
The problem is that Chrome is dominant among browsers🥇

**Comment by Ponant** [comment]
You mean this
https://github.com/aspnet/AspNetCore/blob/574be0d22c1678ed5f6db990aec78b4db587b267/src/Http/Http.Abstractions/src/HostString.cs

**Comment by Ponant** [comment]
Actually that is better
https://docs.microsoft.com/en-us/dotnet/api/system.globalization.idnmapping?view=netframework-4.7.2
it seems to work but one needs to isolate the domain name. It starts getting messy if there are several @ symbols in the email address.

**Comment by blowdart** [comment]
Chrome munges to the left of the @? Aggh. 

The original example, bernhard@xn--gnter-kva.com is arguable right, because it's just the domain name that got IDNified.

**Comment by Ponant** [comment]
1) günter@gmail.com is rejected by Chrome, FF, Edge
2) hello@günter.com gets
-encoded by chrome
-sent raw by FF
-refused by edge

**Comment by Ponant** [comment]
The IDN makes things more complicated as the methods throw if the input on the server is ascii or not and this input will depend on the browser's behavior. I don't really like this pass especially that there are phishing attacks with punycode domains.

**Comment by Ponant** [comment]
@PinpointTownes has a good point about mobile experience with type=text but having that in asp-for will avoid the complexity discussed above.
One can further make the system more robust by having an "AllowedEmailCharacters" as a property in identityoptions together with the validation similar to "AllowedUserNameCharacters" that can be tweaked by the dev who wishes to include or exclude non ascii text.
In my case actually I don't want non-ascii text but I was testing my attributes and got this bug.

**Comment by blowdart** [comment]
Well edge will match Chrome soon enough ...

We did merge a new feature in 3.0 for having [different username and email normalizers](https://github.com/aspnet/AspNetCore/pull/8412) - an email normalizer could reverse IDN on the domain side.

@HaoK do normalizers run before validators?

**Comment by HaoK** [comment]
Normalizes have nothing to do with the actual name field. It’s only for dB lookups

**Comment by blowdart** [comment]
Darn it.

**Comment by Ponant** [comment]
@blowdart in the end I don't think that is your issue :) I mean it would make more sense that this is something at the MVC level, where strings will be puny-decoded to counter chrome's behavior. Having` <input type="text" asp-for="Input.Email"/>` fixes this at the moment.

**Comment by blowdart** [comment]
Except switching it back makes for a worse end-user during input, and is semantically dubious.

**Comment by Ponant** [comment]
I agree, this is a workaround. But in the end I owuld find it more natural if the conversion is made at the action level of an mvc page or handler before entering a usermanager or the identity methods.

**Comment by Ponant** [comment]
ok my understanding now is that google chose to punycode to avoid homograph attacks, e.g.
here
https://аррӏе.com
sounds like apple? check on FF and Chrome and Edge and only FF will show "аррӏе.com" in the url bar.

**Comment by Ponant** [comment]
https://github.com/dotnet/corefx/blob/b8654425442fef4ba4b58510be53d9859887be1a/src/System.Private.Uri/src/System/DomainNameHelper.cs
for the record as I am working on this. This method uses unsafe keyword.

**Comment by Ponant** [comment]
I find it tricky to find a solution that remains consistent. For instance if you allow user to enter an email or url and he enters say http://faß.de . This link in FF will display as is and route to it. But in Chrome and Edge http://faß.de will redirect to https://www.bayern-fass.de/ . WTF can we do with this?

**Comment by blowdart** [comment]
*sigh* Chrome. I think this is going to take thought, which certainly rules it out for the next preview.

**Comment by Ponant** [comment]
Still dwelling upon this issue. I found out that when client side validation is enabled we get some weird behavior:

**For a@ü.com
Client side validation enabled**
FF refuses the post with error "Please enter a valid email address"
Edge same as FF
Chrome allows the post and sends you the punycode a@xn--tda.com

**For a@ü.com
Client side validation disabled**
FF allows the post and sends the raw value a@ü.com
Edge still silly, it refuses the post and displays its default error "You must enter a valid email address"
Chrome allows the post and sends you the punycode a@xn--tda.com

In short, Chrome doe snot care about client side validation, whereas FF will work only if client side val is disabled, Edge refuses unicode altogether. Two problems:
1) Edge needs to fix this and accept unicode
2) The js validation scripts do not like unicode, so these need to be adapted or else FF needs to punycode the entries that are evaluated by jquery validation+unobstrusive.

Any thought?
